### DISCUSSION TITLE
Modified plugins not used 1st time

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#73] [Q&A] [answered]
I seem to be having a bizarre problem with plugins that are only used during a --burn operation not running the most recent version even though I run both --customize and --burn steps.  If I rerun the --customize and --burn steps a second time, the modified plugin finally gets executed.

I believe this sequence is reproducible:

- Modify and save (in my local setup directory) a plugin script.
- Run --customize which does not use that script.
- Run --burn referencing my plugin by full pathname to my local setup directory.
- Observe that the plugin which ran was the previous version.
- Run --customize again (same command).
- Run --burn again (same command).
- Find that the desired updated plugin was run at the appropriate point.

**Comment by gitbls** [comment]
hmm... just to confirm, are you using `--bupdate plugin` on your burn command or not? By default, plugins are NOT updated on a burn unless `--bupdate plugin` is used.

**Reply by dshaw619** [reply]
Afraid I missed that new switch in v7.10.  Should have paid more attention to the change log.
Just to clarify, the correct syntax would be "--bupdate plugin"?  i.e. not the name of any specific plugin?

The Command Details page says "Best to not include it always unless you know what you're doing".  What would be the downsides of doing so?  

p.s. I'd suggest a reference to this be included on the Programming Plugins page.

**Comment by gitbls** [comment]
Thinking was that you would want control over it, so either make 'auto-update' or 'update-by-command-only' the default. I opted for 'by-command-only' on the principle of least suprise. Didn't think anyone would get surprised the other way 🤔

Adding it to the docs: Noted

Edit: I need to revisit this in more detail. Stay tuned.

**Comment by gitbls** [comment, marked as answer]
Here's what I've concluded, and am implementing for the `--burn` command WRT plugins:

- If `--bupdate plugin` is not specified, plugins on the command line specified with only a name (e.g., knockd [in plugins], or myplugin [in local-plugins]) will be checked at burn time. If there is a newer version on the host, a %message will be issued. Plugins referenced on the command line that live in /usr/local/sdm/{local-plugins,plugins} will cause a check between the version on the host and the version on the freshly-burned output
- As above, but if `--bupdate plugin` IS specified, an updated plugin on the host will be updated onto the burned output in /usr/local/sdm/local-plugins. So, if you modify an sdm-supplied plugin in `plugins`, the updated version will be in `local-plugins` and used before the sdm-supplied version, since all sdm code referencing plugins search local-plugins before plugins.
- OTOH, if `--plugin /path/to/someplugin` is specified, the plugin will ALWAYS be updated onto the burned output and placed in /usr/local/sdm/local-plugins. There is no need to specify `--bupdate plugin` in this case. Thinking being that you've decided that the /path/to/someplugin should be written/used in the burned output and referencing it explicitly forces a `--bupdate plugin` for plugins thus referenced.

In other words, to push a brand new plugin into the burned output (that's not in the source IMG), use /path/to/someplugin to force it to be copied in .

Thoughts?

**Reply by dshaw619** [reply]
Sounds great.  Just to make sure I understand completely, if one specifies `--plugin /path/to/someplugin` AND `--bupdate plugin` (as is working for me now; I'll plan to take out the `--bupdate plugin` after the update is released, but one never knows) the behavior will be the same as if `--bupdate plugin` had not been specified?

**Reply by gitbls** [reply]
Exactly.

**Comment by gitbls** [comment]
This is fixed in V7.11

**Reply by dshaw619** [reply]
Doesn't seem to be working correctly (or I've done something really wrong).

- Updated to v7.11 over my existing installation (by running the command at https://github.com/gitbls/sdm#install-sdm).
- Removed the `--bupdate plugin` switch from my --burn script (which was working previously).
- Fixed an error in one of my plugin scripts that is only called during a --burn.
- Ran my --customize and --burn scripts.
- Saw (in the terminal log) that the previous version (3/19) of the plugin had been run.
- Reran --burn script with `--bupdate plugin` switch put back.
- Saw (in the terminal log) that the previous version (3/19) of the plugin had still been run.
- Reran both --customize and --burn (with `--bupdate plugin` switch) scripts.
- The previous version (3/19) of the plugin was still the one that ran.

This shows my plugin, updated today:
```
doug@dv6-Notebook-PC:~$ ls -l ~/setupdir/configure-IncrementalBackup-plugin.sh
-rwxrwxr-x 1 doug doug 4435 Mar 20 12:13 /home/doug/setupdir/configure-IncrementalBackup-plugin.sh
```
and this is the copy in /usr/local/sdm/local-plugins (after the last --customize and --burn above; was same after each burn):
```
doug@dv6-Notebook-PC:~$ ls -l /usr/local/sdm/local-plugins/configure-IncrementalBackup-plugin.sh
-rwxr-xr-x 1 root root 4431 Mar 19 16:03 /usr/local/sdm/local-plugins/configure-IncrementalBackup-plugin.sh
```
setupdir is a link to my working subdirectory:
```
doug@dv6-Notebook-PC:~$ ls -l setupdir
lrwxrwxrwx 1 doug doug 64 Mar  4 17:38 setupdir -> /home/doug/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye
```
And here is my --burn script:
```
#!/bin/bash
# sdm-burn.sh
# Bash file to run sdm and burn an RPi OS image.

setupdir=$HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye    # location of my setup files

echo
date
echo
echo "*Beginning sdm burn"
echo

sudo /usr/local/sdm/sdm --burn /dev/sdd \
--ddsw bs=16M \
--hostname KeyLime \
--plugin burnpwd:"user=des|method=prompt" \
--redact \
--plugin vnc:"realvnc=1280x720" \
--plugin $setupdir/configure-VNC-plugin.sh \
--plugin $setupdir/install-MQTT-plugin.sh:"ROLE=broker" \
--plugin $setupdir/install-Healthcheck-plugin.sh \
--plugin $setupdir/configure-IncrementalBackup-plugin.sh \
--expand-root \
--bupdate plugin \
--plugin-debug \
~/Downloads/bullseye-armhf-customized.img

I realize I shouldn't use "setupdir" as a link to a subdirectory and a parameter within the --burn (and --configure) script, but sdm seems to successfully find all the plugins and other configuration files I reference this way.

**Reply by gitbls** [reply]
Well, I just did the following test:

Customized an uncustomized IMG and installed the plugin `plugdebug` from local-plugins:
```
2023-03-20 17:49:05 > Command Line: /usr/local/bin/sdm --customize --nouser --nowpa --poptions 
2023-03-20 17:49:05   noupdate,noupgrade,noautoremove --plugin plugdebug 2023-02-21-raspios-bullseye-arm64-lite.img
```
The plugin printed:
```
2023-03-20 17:49:06 > Run Plugins Phase '1'
2023-03-20 17:49:06 > Run Plugin 'plugdebug' (/usr/local/sdm/local-plugins/plugdebug) Phase 1 with no arguments
2023-03-20 17:49:07 * Plugin plugdebug: Start Phase 1
2023-03-20 17:49:07 %% DEBUG 1:0
2023-03-20 17:49:07 * Plugin plugdebug: Complete Phase 1
```
I then burned it to an SD card:
```
2023-03-20 17:50:27 > Burn command line: /usr/local/bin/sdm --burn /dev/sdc --plugin /ssd/work/plugdebug 2023-02-21-raspios-bullseye-arm64-lite.img
```
this should pull an updated plugdebug from that directory. I edited the file to change ":0" to ":NEW". Here's what the log has:
```
2023-03-20 17:51:15 > Run Plugins Phase '1'
2023-03-20 17:51:15 > Run Plugin 'plugdebug' (/usr/local/sdm/local-plugins/plugdebug) Phase 1 with no arguments
2023-03-20 17:51:15 * Plugin plugdebug: Start Phase 1
2023-03-20 17:51:15 %% DEBUG 1:NEW
2023-03-20 17:51:15 * Plugin plugdebug: Complete Phase 1
```
So, it did exactly what it should.

Can you provide a simplified case, independent of your infrastructure, so I can replicate it here? Hint: My plugdebug script is precisely sdm-plugin-template with a few additional printouts in it.

**Reply by dshaw619** [reply]
I'll see if I can reproduce in a greatly simplified case.  That was my planned next step.

**Reply by dshaw619** [reply]
I ran a test with greatly reduced scripts and couldn't reproduce the behavior, so I guess the problem must be in something I changed to my scripts.  I'll dig further.

**Reply by gitbls** [reply]
That's good to hear, I guess. Here's what the burn-time plugin update handling does (in case it is helpful for you):

- **If a path was specified on the plugin name:** Copy it into the burned IMG if it's newer than the one in local-plugins in the burned IMG, replacing the one that it found. Always copies if plugin not in local-plugins,, and always copies to local-plugins,
- **If a path was not specified on the plugin name** and the one on the host is newer, update the plugin if `--bupdate plugin` specified, otherwise just log that the host copy is newer

In reviewing the plugin handling I discovered that I was a bit too aggressive in unhooking plugin checking during command processing/arg checking, so will be updating that tomorrow. At the moment, I believe an unknown plugin won't be reported until the burn has completed. Sad and wrong. Thanks for causing me to find it 😄

**Reply by gitbls** [reply]
version 7.11.1 corrects the plugin existence checking at the start of burn command processing, instead of waiting until the disk has been burned.

**Reply by dshaw619** [reply]
I reinstalled sdm over the existing installation at about 13:00 PDT (I think that is after any code mods you made.  At the time I didn't see a changelog entry for v7.11.1, but do now).

I found that the scripts that were failing to be updated had old versions in /usr/local/sdm/local-plugins.  These must have been copied there by an earlier iteration of sdm because I am always using explicit references to all my plugins.

I deleted all the files in /usr/local/sdm/local-plugins and now any plugins I update run correctly.  It seems that the plugins that are used in my --customize script get copied into /usr/local/sdm/local-plugins, but those that are only used in the --burn script don't.

As a further experiment I copied the minimal plugin (call it v17-41) to /usr/local/sdm/local-plugins, then modified the plugin in my set up directory (now v18-50).  I ran just my --burn script (getting lazy after being at this for several hours) and found that the modified (v18-50) was correctly run.  

Then I modified the minimal plugin again (v19-20).  I didn't touch the /usr/local/sdm/local-plugins directory (it still contains v17-41).  Ran both my --customize and --burn scripts and the old v17-41 plugin ran.

So it seems that the scenario that causes problems is:
- old version of plugin in /usr/local/sdm/local-plugins for whatever reason
- modify the plugin
- run --customize (not referencing the plugin) and --burn (calling the plugin by explicit reference to filename)
- old version in /usr/local/sdm/local-plugins is erroneously run during the --burn

**Reply by gitbls** [reply]
My thinking with plugins WRT burn is to NOT update /usr/local/sdm/local-plugins because a burn doesn't do anything else to the current system, it just writes the burn disk. I still think that's the right solution. local-plugins WILL be updated on a customize, though.

As far as your scenario that causes problems, I have a couple of questions:

- When you say "old version of plugin in local-plugins" and then "modify the plugin"...the plugin in WHICH directory gets modified?
- When you run customize, the copy that's in local-plugins should get copied into the IMG, because that's what sdm does when you do a customize (regardless of whether you reference the plugin or not)
- When you burn, if you use `--plugin /full/path/to/plugin` the updated plugin should (and based on my tests, does) get copied into the IMG. Can you please verify that you're using a full path to the plugin?

How are you determining that the old version is being run? If you have one in local-plugins, one in the IMG (which in theory is the same as local-plugins, and gets copied to the burned disk), and one `--plugin /full/path/to/plugin` there are 3 copies of the plugin. How are you convincing yourself that the one in local-plugins is being run? And, if it is the one in local-plugins, it will be the one that's on the burned disk in local-plugins, right?

**Reply by gitbls** [reply]
I just did the following:

- Old version of plugin in /usr/local/sdm/local-plugins on the host
- Run customize (not invoking the plugin)
- Edit /fpt/plugin (/full/path/to/plugin)
- Did burn which included `--plugin /fpt/plugin`

The updated plugin in /fpt/plugin ended up on the burned disk, and was correctly run for all 3 phases after the burn completed.

I think this is exactly the scenario you described, although not 100% sure due to a couple of things you mentioned not being 100% clear

If there's a difference between what you did and what I did, please identify it!

**Reply by dshaw619** [reply]
Frankly, I really am not found of configuration files anywhere that isn't explicitly specified getting used.  It makes some things easy in the short term, but down the line when one has been away from a project for a few months it can be difficult to remember where such things might be 'hidden'.  This is why I explicitly use full paths to all my plugins.  e.g. `--plugin $setupdir/minimal-plugin.sh \`  The variable setupdir was specified at the beginning of the script that I use to run --burn or --customize as `setupdir=$HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye`

Any time I mention creating or modifying a script it would be in my set up directory ($HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye).

In trying to sort all this out, I have put `plugin_dbgprint` statements in the plugins I'm testing in order to determine which version runs. 

The only think I see different about your test is that you ran customize before editing the plugin, but I can't see how that could possibly cause any difference.

I will go back again and see if I can reproduce with bare, minimum scripts.

**Reply by dshaw619** [reply]
I was able to reproduce with greatly reduced scripts.

`/usr/local/sdm/local-plugins` on my laptop contains an older version (17-41) of the plugin.   (the 2nd line was leftover from an even earlier version.  I moved the dbgprint to the main body of the plugin so it would show up in all phases and didn't delete the dbgprint I had originally put in phase 1).
```
doug@dv6-Notebook-PC:~/sdir$ cat /usr/local/sdm/local-plugins/minimal-plugin.sh | grep dbg
plugin_dbgprint "v2023-03-21_17-41"
    plugin_dbgprint "Minimal plugin; v2023-03-21_16-08"
```
sdir is a link to my setup working directory and contains an updated version (21-34) of the plugin.
```
doug@dv6-Notebook-PC:~/sdir$ cat minimal-plugin.sh | grep dbg
plugin_dbgprint "v2023-03-21_21-34"
```
I ran significantly reduced versions of my --customize and --burn scripts.  Only plugin in either is minimal-plugin.sh run during --burn.

Before exiting the --customize, I verified that the image contained the old (17-41) plugin.
```
root@dv6-Notebook-PC:/# cat /usr/local/sdm/local-plugins/minimal-plugin.sh | grep dbg
plugin_dbgprint "v2023-03-21_17-41"
    plugin_dbgprint "Minimal plugin; v2023-03-21_16-08"
```
The plugin switch in the --burn script is `--plugin $setupdir/minimal-plugin.sh \`
setupdir is defined in the --burn script as `setupdir=$HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye`

The older version (17-41) of the plugin was run during the --burn as seen in this excerpt from the terminal:
```
> Run Plugins Phase '1'
> Run Plugin 'minimal-plugin.sh' (/usr/local/sdm/local-plugins/minimal-plugin.sh) Phase 1 with 
  no arguments
D!Plugin minimal-plugin.sh: v2023-03-21_17-41
* Plugin minimal-plugin.sh: Start Phase 1
D!Plugin minimal-plugin.sh: Minimal plugin; v2023-03-21_16-08
* Plugin minimal-plugin.sh: Complete Phase 1
```

**Reply by dshaw619** [reply]
As a further test, I deleted `/usr/local/sdm/local-plugins/minimal-plugin.sh`  No other changes.  Reran my minimal --customize and --burn scripts and  the correct new plugin (21-34) ran during the burn.

I will send you and email with the scripts and terminal logs in case there is anything there that you want to look at.

**Reply by gitbls** [reply]
Amusingly, you nailed it: `The only think I see different about your test is that you ran customize before editing the plugin, but I can't see how that could possibly cause any difference.`

This was precisely the problem.  sdm was copying files (including itself) into the IMG at customization time using `cp`, which changes the file date/time. Since you modified your /fpt/plugin before doing the customize, it appeared to be older than the older plugin, which was copied into the IMG but given a newer date/time.

V7.12 updates all the appropriate `cp` commands in sdm and plugins to use `cp -a`. 

Thanks for your perseverance on this!

**Reply by dshaw619** [reply]
Looks good!

- Updated sdm to v7.12.
- With older version of minimal plugin in `/usr/local/sdm/local-plugins`, I ran my minimal test --customize and --burn scripts (only plugin referenced was a newer copy of the minimal plugin in my working directly called by full path name during the --burn.  Newer version ran as it should.
- Modified the copy of the minimal plugin in my working directory (leaving the old version still in `/usr/local/sdm/local-plugins`).  Ran my full --customize and --burn scripts.  Newest version of the minimal plugin (referenced by full path name during the --burn only) ran as it should.
- Deleted the old version of the minimal plugin in `/usr/local/sdm/local-plugins`.  Modified the copy of the minimal plugin in my working directory.  Ran my full --customize and --burn scripts.  Newest version of the minimal plugin (referenced by full path name during the --burn only) ran as it should.

Thanks so much, gitbls, for your diligence in following through on this and not just brushing it off with "can't replicate" as so many do.

**Reply by gitbls** [reply]
Great! Thanks for persevering on this. I'm pretty excited about plugins, so wanted to get this working as it should.

> Thanks so much, gitbls, for your diligence in following through on this and not just brushing it off with "can't replicate" as so many do.

I appreciate the kind words, and thank you for helping sort this out!
